tests/test_optimize.py

```diff
@@ -671,3 +671,65 @@ def test_update_stepsize(self):
         eta2 = 0.1
         opt.update_stepsize(eta2)
         assert opt._stepsize == eta2
+
+    @pytest.mark.parametrize(
+        "opt",
+        [
+            GradientDescentOptimizer(stepsize),
+            MomentumOptimizer(stepsize, momentum=gamma),
+            NesterovMomentumOptimizer(stepsize, momentum=gamma),
+            AdagradOptimizer(stepsize),
+            RMSPropOptimizer(stepsize, decay=gamma),
+            AdamOptimizer(stepsize, beta1=gamma, beta2=delta),
+            RotosolveOptimizer(),
+        ],
+    )
+    class Test_over_opts:
+        def test_kwargs(self, opt, tol):
+            """Test that the keywords get passed and alter the function"""
+
+            def kwargs_func(x, c=1.0):
+                return (x - c) ** 2
+
+            x_new_one = opt.step(kwargs_func, 1.0, c=1.0)
+            x_new_two = opt.step(kwargs_func, 1.0, c=2.0)
+
+            if getattr(opt, "reset", None):
+                opt.reset()
+
+            assert x_new_one != pytest.approx(x_new_two, abs=tol)
```

Comment:
Nice test!

Perhaps we can be a bit more explicit though in checking that the cost function was called with the correct arguments. The `mocker` pytest fixture can help us do this:

```python
def test_kwargs(self, opt, tol, mocker):
    """Test that the keywords get passed and alter the function"""

    class Cost:
        @staticmethod
        def kwargs_func(x, c=1.0):
            return (x - c) ** 2

    cost = Cost()
    spy = mocker.spy(cost, "kwargs_func")

    x_new_one = opt.step(cost.kwargs_func, 1.0, c=1.0)
    assert spy.call_args_list[-1][1]["c"] == 1.0

    x_new_two = opt.step(cost.kwargs_func, 1.0, c=2.0)
    assert spy.call_args_list[-1][1]["c"] == 2.0

    if getattr(opt, "reset", None):
        opt.reset()

    assert x_new_one != pytest.approx(x_new_two, abs=tol)
```

`mocker.spy` creates a new `Spy` object, that we can use to check the arguments/keyword arguments it was called with.

The reason I wrapped the cost function in a class is that I suddenly realised I don't know how to spy on a pure function 🤔 @antalszava @thisac do you guys have any idea?

Reply:
Hmm, not quite sure actually! It seems to be good when there's a module object to pass as the 1st argument, so e.g., `mocker.spy(qml, "about")` should be fine (see https://github.com/pytest-dev/pytest-mock/issues/97#issuecomment-355250778).